sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * A thread-safe manager for [[CatalogPlugin]]s. It tracks all the registered catalogs, and allow
+ * the caller to look up a catalog by name.
+ */
+class CatalogManager(conf: SQLConf) extends Logging {
+
+  private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
+
+  def catalog(name: String): CatalogPlugin = synchronized {
+    catalogs.getOrElseUpdate(name, Catalogs.load(name, conf))
+  }
+
+  def defaultCatalog: Option[CatalogPlugin] = {
+    conf.defaultV2Catalog.flatMap { catalogName =>
+      try {
+        Some(catalog(catalogName))
+      } catch {
+        case NonFatal(e) =>
+          logError(s"Cannot load default v2 catalog: $catalogName", e)
+          None
+      }
+    }
+  }
+
+  def v2SessionCatalog: Option[CatalogPlugin] = {
+    try {
+      Some(catalog(CatalogManager.SESSION_CATALOG_NAME))
+    } catch {
+      case NonFatal(e) =>
+        logError("Cannot load v2 session catalog", e)
+        None
+    }
+  }
+
+  private def getDefaultNamespace(c: CatalogPlugin) = c match {
+    case c: SupportsNamespaces => c.defaultNamespace()
+    case _ => Array.empty[String]
+  }
+
+  private var _currentNamespace = {
+    // The builtin catalog use "default" as the default database.
+    defaultCatalog.map(getDefaultNamespace).getOrElse(Array("default"))
```

Comment:
@cloud-fan is this the right behavior? Since `CatalogManager` is created in `Analyzer`, unless you set the default catalog (via `spark.sql.default.catalog`) when you create the `SparkSession`, this will be set to `"default"`.

For #25247, you cannot do the following:
```Scala
// Assume that 'testcat' catalog implements SupportsNamespaces and returns Array() for the defaultNamespace.
spark.conf.set("spark.sql.default.catalog", "testcat")

// Now the following will use `default` as the default namespace instead of ``.
spark.sql("SHOW TABLES")
```

But, I see SQL queries where "spark.sql.default.catalog" is set/used without creating a new `SparkSession` here: https://github.com/apache/spark/blob/c4257b18a1efa82d98c8fa9911fa3f3dd95a72b4/sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala#L203

Reply:
Maybe, the test case needs to be updated to create a new `SparkSession` after setting `spark.sql.default.catalog`?

Reply:
Good catch! The `currentNamespace` and `currentCatalog` are not used anywhere and I was planning to add tests when I implement switching current namespace/catalog. Let me add some simple tests to reflect runtime config changes first.

Reply:
Thanks for finding this @imback82!